includes/class-wc-stripe-database-cache.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+defined( 'ABSPATH' ) || exit; // block direct access.
+
+/**
+ * Class WC_Stripe_Database_Cache
+ */
+
+/**
+ * A class for caching data as an option in the database.
+ */
+class WC_Stripe_Database_Cache {
+
+	/**
+	 * In-memory cache for the duration of a single request.
+	 *
+	 * This is used to avoid multiple database reads for the same data and as a backstop in case the database write fails,
+	 * thus ensuring the cache generator is not called multiple times (which would mean multiple API calls).
+	 *
+	 * @var array
+	 */
+	private static $in_memory_cache = [];
+
+	/**
+	 * Class constructor.
+	 */
+	private function __construct() {
+	}
+
+	/**
+	 * Stores a value in the cache.
+	 *
+	 * @param string $key  The key to store the value under.
+	 * @param mixed  $data The value to store.
+	 * @param int    $ttl  The TTL of the cache. Dafault 1 hour.
+	 *
+	 * @return void
+	 */
+	public static function set( $key, $data, $ttl = HOUR_IN_SECONDS ) {
+		self::write_to_cache( $key, $data, $ttl );
+	}
+
+	/**
+	 * Gets a value from the cache.
+	 *
+	 * @param string $key The key to look for.
+	 *
+	 * @return mixed|null The cache contents. NULL if the cache value is expired or missing.
+	 */
+	public static function get( $key ) {
+		$cache_contents = self::get_from_cache( $key );
+		if ( is_array( $cache_contents ) && array_key_exists( 'data', $cache_contents ) ) {
+			if ( self::is_expired( $key, $cache_contents ) ) {
+				return null;
+			}
+
+			return $cache_contents['data'];
+		}
+
+		return null;
+	}
+
+	/**
+	 * Deletes a value from the cache.
+	 *
+	 * @param string $key The key to delete.
+	 *
+	 * @return void
+	 */
+	public static function delete( $key ) {
+		// Remove from the in-memory cache.
+		unset( self::$in_memory_cache[ $key ] );
+
+		// Remove from the DB cache.
+		if ( delete_option( $key ) ) {
+			// Clear the WP object cache to ensure the new data is fetched by other processes.
+			wp_cache_delete( $key, 'options' );
+		}
+	}
+
+	/**
+	 * Wraps the data in the cache metadata and stores it.
+	 *
+	 * @param string  $key     The key to store the data under.
+	 * @param mixed   $data    The data to store.
+	 * @param int     $ttl     The TTL of the cache.
+	 *
+	 * @return void
+	 */
+	private static function write_to_cache( $key, $data, $ttl ) {
+		// Add the  data and expiry time to the array we're caching.
+		$cache_contents            = [];
+		$cache_contents['data']    = $data;
+		$cache_contents['updated'] = time();
+		$cache_contents['ttl']     = $ttl;
```

Comment:
Non-blocking nit: using multiple assignments into the array feels noisy to me, and makes me read a bit harder to fully get that we're writing a single array and what shape it has. I would find it easier to set the contents as a single array, but that's truly a non-blocking stylistic preference.
```suggestion
		$cache_contents            = [
			'data'    => $data,
			'ttl'     => $ttl,
			'updated' => time(),
		];
```